### DISCUSSION TITLE
Can I ( or Should I ) Load Nested Assets into Flat Bundles?

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#1745] [Q&A]
I want to start a discussion on how we might be able to solve a real pain point I have with the current Bevy asset system: the need to ergonomically load nested assets into a flat bundle structure. Let's walk though a simple example.

Say I have two asset types: `SpriteSheet`, and `Sprite`. `Sprite` is an asset that can be loaded from `.png` files, and `SpriteSheet` is an asset that can be loaded from `.spritesheet.yml` files.

Now let's say I have these two assets:

**`player.spritesheet.yml`**:

```yaml
sprite: player.png
grid_size: [16, 16]
```

**`player.png`**: A sprite sheet image with a 16x16

Now, in my code I want to be able to spawn a bundle that looks like this:

```rust
#[derive(Bundle, Default)]
struct SpriteSheetBundle {
    image: Handle<Sprite>,
    sprite_sheet: Handle<SpriteSheet>,
}
```

The issue here, is that because we have a nested structure where `player.spritesheet.yml` references `player.png`, there is no way to load the assets into the flat structure of the `SpriteSheetBundle`. We've got sort of a fundamental disconnect between wanting to have a nice nested asset structure, but needing to load into the flat structure that makes sense to the ECS.

---

I feel like my desire for nested assets on disk feels right from a user experience standpoint, I want to be able to say "spawn a sprite sheet!" not "load a sprite and load a spritesheet and add them to an entity", but it feels contrary to the ECS design and I'm not sure if it's right to try to marry them or to force my assets into a flat structure. 🤷‍♂️

Ideally I really wanted something like this:

```rust
commands
        .spawn()
        .insert_bundle(
            asset_server.load_bundle::<SpriteSheetBundle>("player.spritesheet.yml"),
        )
```

But maybe that doesn't actually make sense.

Anyway, just opening this discussion to see if anybody has any perspectives on this.

**Comment by zicklag** [comment]
I'm thinking maybe what I'm actually looking for are Scenes. In my use-case I'm going for an entirely asset-driven approach to the game, where everything except logic ( and after I figure out scripting, even some of the logic ) is loaded entirely through assets so that you can make different games without having to re-compile any Rust.

So what I'm wanting is to be able to define whole sets of component information in an asset, pretty much just like scenes. The existing scene setup is just way more verbose than I want it to be for exposing to users for modding, etc. Maybe I'll experiment with a simpler scene system that isn't as verbose.

**Comment by mockersf** [comment]
Could you take a look at the [GLTF asset loader](https://github.com/bevyengine/bevy/blob/main/crates/bevy_gltf/src/loader.rs)? It feels like it has a lot of similarity to what you want.

For example, it is loading another asset (an image) from within the loader, and getting it's handle:
https://github.com/bevyengine/bevy/blob/1d7196da4fff00bbc15fe75a04849de915f587d5/crates/bevy_gltf/src/loader.rs#L263-L270


Or it's adding a loaded asset as a labeled asset from the main handle:
https://github.com/bevyengine/bevy/blob/1d7196da4fff00bbc15fe75a04849de915f587d5/crates/bevy_gltf/src/loader.rs#L201

Labeled assets can be reached from the asset server by adding `#label` after the file name:
https://github.com/bevyengine/bevy/blob/1d7196da4fff00bbc15fe75a04849de915f587d5/examples/asset/asset_loading.rs#L20

**Reply by zicklag** [reply]
Hmm, that is alot like I was trying to do. That might give me some hints. I also noticed that in the glTF example, the glTF file is spawned as a scene: https://github.com/bevyengine/bevy/blob/1d7196da4fff00bbc15fe75a04849de915f587d5/examples/3d/load_gltf.rs#L12

That might be the trick, then. To make an asset that you can load that _produces_ a scene that can be spawned. That way I could make a simple YAML file that has an asset loader that produces a bevy scene ( like the glTF loader does ), that can then be spawned with a single command!

That might work. I'll try it out. Thanks!
